intellij-plugin-structure/structure-intellij/src/main/java/com/jetbrains/plugin/structure/intellij/plugin/OptionalDependencyResolver.kt

```diff
@@ -0,0 +1,62 @@
+package com.jetbrains.plugin.structure.intellij.plugin
+
+import com.jetbrains.plugin.structure.intellij.plugin.dependencies.DependencyChain
+import com.jetbrains.plugin.structure.intellij.problems.OptionalDependencyDescriptorResolutionProblem
+import com.jetbrains.plugin.structure.intellij.problems.PluginCreationResultResolver
+import com.jetbrains.plugin.structure.intellij.resources.ResourceResolver
+import java.nio.file.Path
+
+internal class OptionalDependencyResolver(private val pluginLoader: PluginLoader) {
+
+  fun resolveOptionalDependencies(
+    plugin: PluginCreator,
+    pluginArtifactPath: Path,
+    resourceResolver: ResourceResolver,
+    problemResolver: PluginCreationResultResolver
+  ) {
+    val dependencyChain = DependencyChain()
+    plugin.resolveOptionalDependencies(pluginArtifactPath, resourceResolver, problemResolver, dependencyChain)
+    dependencyChain.cycles.forEach {
+      plugin.registerOptionalDependenciesConfigurationFilesCycleProblem(it)
+    }
+  }
+
+  /**
+   * @receiver plugin whose optional dependencies are resolved (plugin.xml, then someOptional.xml, ...)
+   */
+  private fun PluginCreator.resolveOptionalDependencies(
+    pluginArtifactPath: Path,
+    resourceResolver: ResourceResolver,
+    problemResolver: PluginCreationResultResolver,
+    dependencyChain: DependencyChain
+  ) {
+    if (!dependencyChain.extend(this)) {
+      return
+    }
+    for ((pluginDependency, configurationFile) in optionalDependenciesConfigFiles) {
+      if (dependencyChain.detectCycle(configurationFile)) {
+        return
```

Comment:
Shouldn't be there `continue` instead of `return` to gather all cycles?

Reply:
Replaced with `continue` and added more unit tests. In addition, fixed a corner-case issue.